src/index.rs

```diff
@@ -332,72 +303,82 @@ impl ResourceIndex {
 
-        let path_buf = CanonicalPathBuf::canonicalize(path)?;
-        let path = path_buf.as_canonical_path();
-
-        return match fs::metadata(path) {
-            Err(_) => {
-                return Err(ArklibError::Path(
-                    "Couldn't to retrieve file metadata".into(),
-                ));
-            }
-            Ok(metadata) => match scan_entry(path, metadata) {
-                Err(_) => {
-                    return Err(ArklibError::Path(
-                        "The path points to a directory or empty file".into(),
-                    ));
-                }
-                Ok(new_entry) => {
-                    let id = new_entry.id;
-
-                    if let Some(nonempty) = self.collisions.get_mut(&id) {
-                        *nonempty += 1;
-                    }
-
-                    let mut added = HashMap::new();
-                    added.insert(path_buf.clone(), id);
-
-                    self.id2path.insert(id, path_buf.clone());
-                    self.path2id.insert(path_buf, new_entry);
+        let path_buf = fs::canonicalize(path)?;
+        let path = path_buf.as_path();
+
+        let metadata = fs::metadata(path).map_err(|e| {
+            ArklibError::Path(format!(
+                "Couldn't to retrieve file metadata: {}",
+                e
+            ))
+        })?;
+        let new_entry = scan_entry(path, metadata)?;
+        let id = new_entry.id;
+        if let Some(nonempty) = self.collisions.get_mut(&id) {
+            *nonempty += 1;
+        }
+        let mut added = HashMap::new();
+        added.insert(path_buf.clone(), id);
+        self.id2path.insert(id, path_buf.clone());
+        self.path2id.insert(path_buf, new_entry);
 
-                    Ok(IndexUpdate {
-                        added,
-                        deleted: HashSet::new(),
-                    })
-                }
-            },
-        };
+        Ok(IndexUpdate {
+            added,
+            deleted: HashSet::new(),
+        })
     }
 
-    // the caller must ensure that:
-    // * the index is up-to-date except this single path
-    // * the path has been indexed before
-    // * the path maps into `old_id`
-    // * the content by the path has been modified
+    /// Updates a single entry in the index with a new resource located at the
+    /// specified path, replacing the old resource associated with the given
+    /// ID.
+    ///
+    /// # Restrictions
+    ///
+    /// The caller must ensure that:
+    /// * the index is up-to-date except for this single path
+    /// * the path has been indexed before
+    /// * the path maps into `old_id`
+    /// * the content by the path has been modified
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the path does not exist, if the path is a directory
+    /// or an empty file, if the index cannot find the specified path, or if
+    /// the content of the path has not been modified.
     pub fn update_one(
```

Comment:
It seems to me, that we should separate this method into `track_deletion` and `track_update` as it was done in #72

https://github.com/ARK-Builders/arklib/blob/index/update-one/src/index.rs#L393
https://github.com/ARK-Builders/arklib/blob/index/update-one/src/index.rs#L416

The idea is that `track` methods serve as performance workaround when end-users know what they are doing. On other hand, `update_all` should be efficient enough to be called all the time. But from API point of view, it `track` methods are also simpler sometimes.

OPEN FOR DISCUSSION

Reply:
Ok so the API will look like:
- `track_addition()`, `track_update()` and `track_deletion()` when the user wants to be specific about the action
- `update_all()` otherwise

This looks like a clean API but I have some questions about other public methods
- what are the use cases for `ResourceIndex::provide()`?
- we can now remove `forget_id()` in favor of `track_deletion()` or they do different tasks?

Reply:
@tareknaser if I remember correctly:
* Function `forget_id` is only utilitary, became private `forget_entry` in the other branch of the codebase. It's dual to `insert_entry`, both are called from `track` methods.
* `ResourceIndex::provide()` is the starting point for any user/app. It combines index loading and updating in single step. This could be done by separate calls, but it seems to me that users will always need updated index before they do something. I can imagine rare use-cases where they need more control, so we should expose separate steps, too.

Reply:
> ResourceIndex::provide() is the starting point for any user/app. It combines index loading and updating in single step.

Got it
Thanks for clearing this out. 

> Name forget_id is confusing now. 


The initial idea for this PR was to rewrite the file `src/index.rs` as it is and then implement the changes to replace #72 
Do you prefer to include all the changes in the current PR?

Reply:
Yeah... It's difficult to keep in mind several PRs because the scope is pretty big.

Let's incorporate useful parts of #72 and unit tests from there into this PR.